src/App.tsx

```diff
@@ -1,61 +1,106 @@
+import { useCallback, useState } from 'react';
 import './App.scss';
 
-// import usersFromServer from './api/users';
-// import todosFromServer from './api/todos';
+import usersFromServer from './api/users';
+import { TodoList } from './components/TodoList';
+import todosFromServer from './api/todos';
+import { Todo, User } from './types';
+
+const defaultFormValue = {
+  title: '',
+  userId: 0,
+};
+
+export const App: React.FC = () => {
+  const [users] = useState<User[]>(usersFromServer);
+  const [todos, setTodos] = useState<Todo[]>(todosFromServer);
+  const [formValue, setFormValue] = useState(defaultFormValue);
+  const [titleTouched, setTitleTouched] = useState(false);
+  const [userTouched, setUserTouched] = useState(false);
+
+  const setFormTouched = (isTouched: boolean) => {
+    setTitleTouched(isTouched);
+    setUserTouched(isTouched);
+  };
+
+  const onSubmitEvent = (event: React.FormEvent<HTMLFormElement>) => {
+    event.preventDefault();
+    if (!formValue.title || formValue.userId === 0) {
+      setFormTouched(true);
+
+      return;
+    }
+
+    const ids = todos.map(todo => todo.id);
+    const newId = Math.max(...ids) + 1;
+    const newTodo: Todo = {
+      id: newId,
+      title: formValue.title,
+      completed: false,
+      userId: formValue.userId,
+    };
+
+    setTodos([...todos, newTodo]);
+    setFormValue(defaultFormValue);
+    setFormTouched(false);
+  };
+
+  const updateFormValue = useCallback((key: string, value: string | number) => {
+    setFormValue(previousValue => ({ ...previousValue, [key]: value }));
+  }, []);
 
-export const App = () => {
   return (
     <div className="App">
       <h1>Add todo form</h1>
 
-      <form action="/api/todos" method="POST">
+      <form action="/api/todos" method="POST" onSubmit={onSubmitEvent}>
         <div className="field">
-          <input type="text" data-cy="titleInput" />
-          <span className="error">Please enter a title</span>
+          <label htmlFor="titleInput">Title: </label>
+          <input
+            type="text"
+            id="titleInput"
+            value={formValue.title}
+            data-cy="titleInput"
+            placeholder="Enter a title"
+            onChange={event => {
+              updateFormValue('title', event.target.value);
+              setTitleTouched(false);
+            }}
```

Comment:
Pls move the function inside `onChange` handler to a separate const and use as a callback here
**PLS DO THE SAME FOR ALL SUCH CASES IN THE PROJECT**